tests/lax_numpy_test.py

```diff
@@ -2998,25 +2998,31 @@ def testArrayCreationWithSharding(self, func, shape, dtype):
       func=[
         lambda dtype, device: jnp.arange(5, dtype=dtype, device=device),
         lambda dtype, device: jnp.eye(5, 6, dtype=dtype, device=device),
+        lambda dtype, device: jnp.linspace(5, 6, 7, dtype=dtype, device=device),
+        lambda dtype, device: jnp.linspace(5, 6, 7, retstep=True, dtype=dtype, device=device),
+        lambda dtype, device: jnp.array([1, 2, 3, 4, 5], dtype=dtype, device=device),
       ],
       dtype=default_dtypes,
   )
-  def testArangeEyeWithDevice(self, func, dtype):
+  def testArangeEyeLinspaceArrayWithDevice(self, func, dtype):
     device = jax.devices()[-1]
-    out = func(dtype=dtype, device=device)
-    self.assertEqual(out.devices(), {device})
+    output = func(dtype=dtype, device=device)
+    jax.tree.map(lambda x: self.assertEqual(x.devices(), {device}), output)
```

Comment:
I think this would be expected to fail in general for the `retstep=True` case, since `delta` goes to the default device. I'd probably just remove it, and go back to the previous simple assertion.

Reply:
I was also thinking about that but it is not failing because under jit:
```
import jax
import jax.numpy as jnp

print(jax.__version__)
out = jnp.linspace(5, 6, 7, retstep=True, device=jax.devices("cpu")[0])
print(out[0].device, out[1].device)

out = jnp.linspace(5, 6, 7, retstep=True, device=jax.devices("cuda")[-1])
print(out[0].device, out[1].device)
```
and the output:
```
0.4.31.dev20240625+72d3c5573
TFRT_CPU_0 TFRT_CPU_0
cuda:1 cuda:1
```

Reply:
I think it's not failing because `devices()[-1]` happens to be the default device within your `pytest` call.

In general this will not be the case.

Reply:
In the above code snippet with the output that I get, I have 2 GPUs and 1 CPU available on the machine. Specifying device as cpu, we can see that `delta` is also on cpu and not on default device cuda:0. Same for cuda:1. Yash (@yashk2810 ) was also confirming in the chat that due to `linspace` is jitted:
```python
@partial(jit, static_argnames=('num', 'endpoint', 'retstep', 'dtype', 'axis', 'device'))
def _linspace(start: ArrayLike, stop: ArrayLike, num: int = 50,
```
the output will be on the right device.

Reply:
Let's put it this way: our API guarantee here is that `out[0]` will be on the specified device. We deliberately make no guarantee about the device of `out[1]`, so our tests should not assert anything about it, or else we're in danger of the test starting to fail sometime in the future.